File: API/0278_Donchian_Width_Mean_Reversion/DonchianWidthMeanReversionStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
	/// <summary>
	/// Donchian Width Mean Reversion Strategy.
	/// This strategy trades based on the mean reversion of the Donchian Channel width.
	/// </summary>
	public class DonchianWidthMeanReversionStrategy : Strategy
	{
		private readonly StrategyParam<int> _donchianPeriod;
		private readonly StrategyParam<int> _lookbackPeriod;
		private readonly StrategyParam<decimal> _deviationMultiplier;
		private readonly StrategyParam<decimal> _stopLossPercent;
		private readonly StrategyParam<DataType> _candleType;
		
		private DonchianChannels _donchian;
		private SimpleMovingAverage _widthAverage;
		private StandardDeviation _widthStdDev;
		
		private decimal _currentWidth;
		private decimal _prevWidth;
		private decimal _prevWidthAverage;
		private decimal _prevWidthStdDev;

		/// <summary>
		/// Donchian Channel period.
		/// </summary>
		public int DonchianPeriod
		{
			get => _donchianPeriod.Value;
			set => _donchianPeriod.Value = value;
		}

		/// <summary>
		/// Lookback period for calculating the average and standard deviation of width.
		/// </summary>
		public int LookbackPeriod
		{
			get => _lookbackPeriod.Value;
			set => _lookbackPeriod.Value = value;
		}

		/// <summary>
		/// Deviation multiplier for mean reversion detection.
		/// </summary>
		public decimal DeviationMultiplier
		{
			get => _deviationMultiplier.Value;
			set => _deviationMultiplier.Value = value;
		}

		/// <summary>
		/// Stop loss percentage.
		/// </summary>
		public decimal StopLossPercent
		{
			get => _stopLossPercent.Value;
			set => _stopLossPercent.Value = value;
		}

		/// <summary>
		/// Candle type.
		/// </summary>
		public DataType CandleType
		{
			get => _candleType.Value;
			set => _candleType.Value = value;
		}

		/// <summary>
		/// Constructor.
		/// </summary>
		public DonchianWidthMeanReversionStrategy()
		{
			_donchianPeriod = Param(nameof(DonchianPeriod), 20)
				.SetDisplay("Donchian Period", "Donchian Channel period", "Donchian")
				.SetCanOptimize(true)
				.SetOptimize(10, 50, 5);

			_lookbackPeriod = Param(nameof(LookbackPeriod), 20)
				.SetDisplay("Lookback Period", "Lookback period for calculating the average and standard deviation of width", "Mean Reversion")
				.SetCanOptimize(true)
				.SetOptimize(10, 50, 5);

			_deviationMultiplier = Param(nameof(DeviationMultiplier), 2.0m)
				.SetDisplay("Deviation Multiplier", "Deviation multiplier for mean reversion detection", "Mean Reversion")
				.SetCanOptimize(true)
				.SetOptimize(1.0m, 3.0m, 0.5m);

			_stopLossPercent = Param(nameof(StopLossPercent), 2.0m)
				.SetDisplay("Stop Loss %", "Stop loss percentage", "Risk Management")
				.SetCanOptimize(true)
				.SetOptimize(1.0m, 5.0m, 0.5m);

			_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
				.SetDisplay("Candle Type", "Candle type for strategy", "General");
		}

		/// <inheritdoc />
		public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		{
			return [(Security, CandleType)];
		}

		/// <inheritdoc />
		protected override void OnStarted(DateTimeOffset time)
		{
			base.OnStarted(time);

			// Initialize indicators
			_donchian = new DonchianChannels { Length = DonchianPeriod };
			_widthAverage = new SimpleMovingAverage { Length = LookbackPeriod };
			_widthStdDev = new StandardDeviation { Length = LookbackPeriod };
			
			// Reset stored values
			_currentWidth = 0;
			_prevWidth = 0;
			_prevWidthAverage = 0;
			_prevWidthStdDev = 0;

			// Create subscription and bind indicators
			var subscription = SubscribeCandles(CandleType);
			subscription
				.BindEx(_donchian, ProcessDonchian)
				.Start();

			// Setup chart visualization if available
			var area = CreateChartArea();
			if (area != null)
			{
				DrawCandles(area, subscription);
				DrawIndicator(area, _donchian);
				DrawOwnTrades(area);
			}
			
			// Start position protection
			StartProtection(
				takeProfit: null,
				stopLoss: new Unit(StopLossPercent, UnitTypes.Percent)
			);
		}

		private void ProcessDonchian(ICandleMessage candle, IIndicatorValue value)
		{
			// Skip unfinished candles
			if (candle.State != CandleStates.Finished)
				return;

			// Check if strategy is ready to trade
			if (!IsFormedAndOnlineAndAllowTrading())
				return;
			
			// Extract upper and lower bands from the indicator value
			var donchianValue = value.GetValue<(decimal upper, decimal middle, decimal lower)>();
			var upperBand = donchianValue.upper;
			var lowerBand = donchianValue.lower;
			
			// Calculate the Donchian channel width
			_currentWidth = upperBand - lowerBand;
			
			// Calculate the average and standard deviation of the width
			var widthAverage = _widthAverage.Process(_currentWidth, candle.ServerTime, candle.State == CandleStates.Finished).ToDecimal();
			var widthStdDev = _widthStdDev.Process(_currentWidth, candle.ServerTime, candle.State == CandleStates.Finished).ToDecimal();
			
			// Skip the first value
			if (_prevWidth == 0)
			{
				_prevWidth = _currentWidth;
				_prevWidthAverage = widthAverage;
				_prevWidthStdDev = widthStdDev;
				return;
			}
			
			// Calculate thresholds
			var narrowThreshold = _prevWidthAverage - _prevWidthStdDev * DeviationMultiplier;
			var wideThreshold = _prevWidthAverage + _prevWidthStdDev * DeviationMultiplier;
			
			// Trading logic:
			// When channel is narrowing (compression), enter long position
			if (_currentWidth < narrowThreshold && _prevWidth >= narrowThreshold && Position == 0)
			{
				BuyMarket(Volume);
				LogInfo($"Donchian channel width compression: {_currentWidth} < {narrowThreshold}. Buying at {candle.ClosePrice}");
			}
			// When channel is widening (expansion), enter short position
			else if (_currentWidth > wideThreshold && _prevWidth <= wideThreshold && Position == 0)
			{
				SellMarket(Volume);
				LogInfo($"Donchian channel width expansion: {_currentWidth} > {wideThreshold}. Selling at {candle.ClosePrice}");
			}
			
			// Exit positions when width returns to average
			else if ((Position > 0 || Position < 0) && 
					 Math.Abs(_currentWidth - _prevWidthAverage) < 0.1m * _prevWidthStdDev &&
					 Math.Abs(_prevWidth - _prevWidthAverage) >= 0.1m * _prevWidthStdDev)
			{
				if (Position > 0)
				{
					SellMarket(Math.Abs(Position));
					LogInfo($"Donchian width returned to average: {_currentWidth} ≈ {_prevWidthAverage}. Closing long position at {candle.ClosePrice}");
				}
				else if (Position < 0)
				{
					BuyMarket(Math.Abs(Position));
					LogInfo($"Donchian width returned to average: {_currentWidth} ≈ {_prevWidthAverage}. Closing short position at {candle.ClosePrice}");
				}
			}
			
			// Store current values for next comparison
			_prevWidth = _currentWidth;
			_prevWidthAverage = widthAverage;
			_prevWidthStdDev = widthStdDev;
		}
	}
}

File: API/0278_Donchian_Width_Mean_Reversion/PY/donchian_width_mean_reversion_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

import math
from System import TimeSpan, Math
from StockSharp.Messages import DataType, Unit, UnitTypes, CandleStates
from StockSharp.Algo.Indicators import DonchianChannels
from StockSharp.Algo.Strategies import Strategy

class donchian_width_mean_reversion_strategy(Strategy):
    """
    Donchian width mean reversion strategy.
    Trades contractions and expansions of Donchian Channel width around its recent average.
    """

    def __init__(self):
        super(donchian_width_mean_reversion_strategy, self).__init__()

        self._donchian_period = self.Param("DonchianPeriod", 20) \
            .SetGreaterThanZero() \
            .SetDisplay("Donchian Period", "Donchian Channel period", "Indicators")

        self._lookback_period = self.Param("LookbackPeriod", 20) \
            .SetGreaterThanZero() \
            .SetDisplay("Lookback Period", "Lookback period for width statistics", "Strategy Parameters")

        self._deviation_multiplier = self.Param("DeviationMultiplier", 1.5) \
            .SetGreaterThanZero() \
            .SetDisplay("Deviation Multiplier", "Deviation multiplier for mean reversion detection", "Strategy Parameters")

        self._stop_loss_percent = self.Param("StopLossPercent", 2.0) \
            .SetGreaterThanZero() \
            .SetDisplay("Stop Loss %", "Stop loss percentage", "Risk Management")

        self._cooldown_bars = self.Param("CooldownBars", 1200) \
            .SetDisplay("Cooldown Bars", "Bars to wait between orders", "Risk Management")

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))) \
            .SetDisplay("Candle Type", "Candle type for strategy", "General")

        self._donchian = None
        self._width_history = None
        self._current_index = 0
        self._filled_count = 0
        self._cooldown = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(donchian_width_mean_reversion_strategy, self).OnReseted()
        self._donchian = None
        lb = int(self._lookback_period.Value)
        self._width_history = [0.0] * lb
        self._current_index = 0
        self._filled_count = 0
        self._cooldown = 0

    def OnStarted(self, time):
        super(donchian_width_mean_reversion_strategy, self).OnStarted(time)

        lb = int(self._lookback_period.Value)
        self._width_history = [0.0] * lb
        self._current_index = 0
        self._filled_count = 0
        self._cooldown = 0

        self._donchian = DonchianChannels()
        self._donchian.Length = int(self._donchian_period.Value)

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.BindEx(self._donchian, self._process_donchian).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, self._donchian)
            self.DrawOwnTrades(area)

        self.StartProtection(Unit(), Unit(self._stop_loss_percent.Value, UnitTypes.Percent))

    def _process_donchian(self, candle, donchian_value):
        if candle.State != CandleStates.Finished:
            return

        if not self._donchian.IsFormed:
            return

        upper_band = donchian_value.UpperBand
        lower_band = donchian_value.LowerBand
        if upper_band is None or lower_band is None:
            return

        upper_val = float(upper_band)
        lower_val = float(lower_band)

        width = upper_val - lower_val

        lb = int(self._lookback_period.Value)
        self._width_history[self._current_index] = width
        self._current_index = (self._current_index + 1) % lb

        if self._filled_count < lb:
            self._filled_count += 1

        if self._filled_count < lb:
            return

        avg_width = 0.0
        for i in range(lb):
            avg_width += self._width_history[i]
        avg_width /= float(lb)

        sum_sq = 0.0
        for i in range(lb):
            diff = self._width_history[i] - avg_width
            sum_sq += diff * diff
        std_width = math.sqrt(sum_sq / float(lb))

        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        if self._cooldown > 0:
            self._cooldown -= 1
            return

        dm = float(self._deviation_multiplier.Value)
        narrow_threshold = avg_width - std_width * dm
        wide_threshold = avg_width + std_width * dm

        if self.Position == 0:
            if width < narrow_threshold:
                self.BuyMarket()
                self._cooldown = int(self._cooldown_bars.Value)
            elif width > wide_threshold:
                self.SellMarket()
                self._cooldown = int(self._cooldown_bars.Value)
        elif self.Position > 0 and width >= avg_width:
            self.SellMarket(Math.Abs(self.Position))
            self._cooldown = int(self._cooldown_bars.Value)
        elif self.Position < 0 and width <= avg_width:
            self.BuyMarket(Math.Abs(self.Position))
            self._cooldown = int(self._cooldown_bars.Value)

    def CreateClone(self):
        return donchian_width_mean_reversion_strategy()
